### PR TITLE
feat: add cloudendpoint to azure flags

### PR DESCRIPTION
**Pull request by faiq**
**What problem does this PR solve?**:
Allows users to pass in cloud endpoint to their packer build #444 

**Which issue(s) does this PR fix?**:

* https://jira.d2iq.com/browse/D2IQ-NUMBER

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/azure/azure.go [hunk 1]
```diff
@@ -61,8 +64,12 @@ func createGalleryImage(
 	description *ImageDescription,
 	location string,
 	subscriptionID string,
+	endpoint arm.Endpoint,
 ) (*armcompute.GalleryImage, error) {
-	galleryImageClient := armcompute.NewGalleryImagesClient(subscriptionID, cred, nil)
+	clientOptions := &arm.ClientOptions{
```
[the remark this answers is not in the document]
Comment (faiq) [reply pkg/azure/azure.go:69]: Makes sense, we can hold until that's merged in.

[the remark this answers is not in the document]
Comment (supershal) [reply pkg/azure/azure.go:69]: Thanks Faiq. I have merged https://github.com/mesosphere/konvoy-image-builder/pull/466

**Comment by supershal** [comment]
Looking at the newer azure SDK API, it seems like you will need both `audience` and `endpoint`. https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/arm/arm.go#L12

May be since there are three fix endpoints, we change `--cloud-endpoint` to enum: AzureChina, AzureGovernment, AzurePublic and manage those endpoints internally in KIB code? wdyt ?

File: pkg/app/build_azure.go [hunk 2]
```diff
@@ -25,6 +27,51 @@ type AzureArgs struct {
 
 	SubscriptionID string
 	TenantID       string
+	CloudEndpoint  *AzureCloudFlag
+}
+
+type AzureCloudFlag struct {
```
Comment (supershal) [inline pkg/app/build_azure.go:33]: niiiiiiice. !!

File: pkg/app/config.go [hunk 3]
```diff
@@ -471,6 +472,13 @@ func MergeAzureUserArgs(config Config, azureArgs *AzureArgs) error {
 		return fmt.Errorf("failed to set %s: %w", PackerAzureInstanceType, err)
 	}
 
+	// packer values
+	// see https://www.packer.io/plugins/builders/azure/arm#cloud_environment_name
+	fmt.Println(azureArgs.CloudEndpoint.String())
```
Comment (supershal) [inline pkg/app/config.go:477]: please remove printing logs here.

File: pkg/app/build_azure.go [hunk 4]
```diff
@@ -49,7 +96,21 @@ func azureCredentials(config Config) (*azure.Credentials, error) {
...
+	var cloudConfig cloud.Configuration
+	switch endpoint {
+	case string(AzureCloudEndpointChina):
+		cloudConfig = cloud.AzureChina
+	case string(AzureCloudEndpointGovernment):
+		cloudConfig = cloud.AzureGovernment
+
+	default:
+		cloudConfig = cloud.AzurePublic
+	}
```
Comment (supershal) [inline pkg/app/build_azure.go:112]: the global variables for cloudconfig does not set the endpoint (https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.1.2/cloud#pkg-variables)
I think you have to set it explicitly here. 
```go
// for AzureChina
cloudConfig.Services[cloud.ResourceManager] = cloud.ServiceConfiguration{
		Audience: "https://management.core.chinacloudapi.cn",
		Endpoint: "https://management.chinacloudapi.cn",
	}

// for AzureGovernment
	cloudconfig.Services[cloud.ResourceManager] = cloud.ServiceConfiguration{
		Audience: "https://management.core.usgovcloudapi.net",
		Endpoint: "https://management.usgovcloudapi.net",
	}
// for AzurePublic
	cloudconfig.Services[cloud.ResourceManager] = cloud.ServiceConfiguration{
		Audience: "https://management.core.windows.net/",
		Endpoint: "https://management.azure.com",
	}
```
ex: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/arm/arm.go#L12

**Review by supershal — CHANGES_REQUESTED** [review]

File: pkg/app/build_azure.go (hunk 4 above)
Comment (faiq) [reply pkg/app/build_azure.go:112]: the line you linked to is an init function that does that sets the [cloud.ResourceManager] field  in the Service for me with the right Audience and Endpoint 
 
When I spew the cloudConfig we see its configured correctly
 ```
 2022/08/24 12:57:26 writing new packer configuration to work/centos-7-1661371046-VMmAq
(cloud.Configuration) {
 ActiveDirectoryAuthorityHost: (string) (len=34) "https://login.microsoftonline.com/",
 Services: (map[cloud.ServiceName]cloud.ServiceConfiguration) (len=1) {
  (cloud.ServiceName) (len=15) "resourceManager": (cloud.ServiceConfiguration) {
   Audience: (string) (len=36) "https://management.core.windows.net/",
   Endpoint: (string) (len=28) "https://management.azure.com"
  }
 }
}
```

**Review by supershal — APPROVED** [review]
Thanks for the improvements.

**Review by dkoshkin — APPROVED** [review]

**Pull request merged**